# Ejercicios evaluables - Matemáticas para la IA

In [ ]:
import numpy as np
import numpy.linalg as linalg
import math
import time

### Ejercicio 1

#### Apartado a)

Se implementa una función que calcule el determinante con la definición de Laplace.

In [ ]:
def determinante_recursivo(matrix):
    n = matrix.shape[0]

    if n != matrix.shape[1]:
        return 'No existe, la matriz no es cuadrada'

    if n == 1:
        return matrix[0, 0]

    det_matrix = 0  # Det inicializado a 0

    for j in range(n):
        cofactor = (-1) ** j * determinante_recursivo(np.delete(np.delete(matrix, 0, 0), j, 1))
        det_matrix += matrix[0, j] * cofactor

    return det_matrix

Matriz de prueba:

In [ ]:
A = np.random.randint(0,9, size=(5,5))
det = determinante_recursivo(A)
print(A, '\n')
print('det(A) =', det)

[[6 4 2 8 7]
 [0 2 0 7 3]
 [5 5 8 4 3]
 [6 0 4 1 4]
 [1 0 8 3 2]] 

det(A) = 1014


#### Apartado b)

Si la matriz es triangular, el determinante será igual al producto de la diagonal. Esto se debe a que basta calcular el determinante por la fila o columna donde son todo ceros menos un elemento, y así sucesivamente. Se crea ahora una función que calcule el determinante de una matriz triangular:

In [ ]:
def determinante_triangular(matrix):

    det = 1

    for i in range(matrix.shape[0]):
        det *= matrix[i,i]

    return det

#### Apartado c)
Si se intercambia una fila por otra fila o una columna por otra columna entonces el determinante queda multiplicado por -1.



Por otro lado, si a una fila (o columna) se le suma otra fila (o columna) multiplicada por un escalar $\alpha$, el determinante no se ve afectado.

#### Apartado d)

In [ ]:
def eliminacion_gauss(A):
    A = A.astype('float64')  #matriz a float
    cont = 0 # Contador que cuenta los cambios de filas

    m, n = A.shape
    for i in range(m):
        # Encuentra el pivote máximo en la columna actual
        pivot_row = np.argmax(np.abs(A[i:, i])) + i

        # Intercambia filas si es necesario
        if pivot_row != i:
            A[[i, pivot_row]] = A[[pivot_row, i]]
            cont += 1

        # Aplica eliminación gaussiana a las filas siguientes
        for j in range(i+1, m):
            factor = A[j, i] / A[i, i]
            A[j, i:] -= factor * A[i, i:]

    A = A.T

    return (A,cont)

matriz_triangular, n_cambio_filas = eliminacion_gauss(A)
print(matriz_triangular,'\n\n')
n_cambio_filas

[[ 6.          0.          0.          0.          0.        ]
 [ 4.         -4.          0.          0.          0.        ]
 [ 2.          2.          7.33333333  0.          0.        ]
 [ 8.         -7.          2.83333333 -8.35227273  0.        ]
 [ 7.         -3.          1.33333333 -5.38636364 -0.68979592]] 




3

#### Apartado e)

Basta realizar a una matriz el método de eliminación de Gauss con pivoteo parcial y luego calcular el determinante a partir del producto de los elementos de la diagonal. Esto es consecuencia de que sumar a una fila otra fila multiplicada por un escalar no altera el determinante. También hay que tener en cuenta que al usar pivotaje parcial estamos intercambiando filas, luego hay que multiplicar por el determinante por $(-1)^j$, siendo $j$ el número total de veces que hemos intercambiado dos filas.

Se implementa a continuación dicha función:

In [ ]:
def determinante_gauss(matrix):
    triang, j = eliminacion_gauss(matrix)
    return determinante_triangular(triang)*(-1)**j

In [ ]:
det = determinante_gauss(A)
print(A, '\n')
print('det(A) =', round(det))

[[6 4 2 8 7]
 [0 2 0 7 3]
 [5 5 8 4 3]
 [6 0 4 1 4]
 [1 0 8 3 2]] 

det(A) = 1014


Observamos como obtenemos el mismo resultado que usando la función *determinante_recursivo*.

#### Apartado f)

El número total de operaciones para realizar eliminación gaussiana sobre una matriz cuadrada de tamaño $n$ es:

$$\frac{n(n+1)}{2}+ \frac{2n^3+3n^2-5n}{6} + \frac{2n^3+3n^2-5n}{6},$$

correspondientes al número de divisiones, número de multiplicaciones y números de restas respectivamente. El número de operaciones para calcular el determinante una vez hecha la eliminación gaussiana es exactamente $n-1$ productos, luego el coste computacional es del orden de $O(n^3)$.

Por otro lado, utilizando la definición de determinante, al ser una función recursiva, el coste computacional se comporta asintóticamente como $O(n!)$.

Todo esto nos indica que para matrices pequeñas ($n$ aproximadamente menor que 6) es mejor usar la definición recursiva, pero para tamaños mayores es mucho más eficiente utilizar el metodo de eliminación de Gauss con pivoteo parcial.

#### Apartado g)

In [ ]:
def comparacion_tiempos():

    for i in range (2,11):
        a=np.random.rand(i,i)

        inicio_1 = time.time()
        determinante_recursivo(a)
        fin_1 = time.time()
        tiempo_1 = (fin_1 - inicio_1)*1e6

        inicio_2 = time.time()
        determinante_gauss(a)
        fin_2 = time.time()
        tiempo_2 = (fin_2 - inicio_2)*1e6

        inicio_3 = time.time()
        np.linalg.det(a)
        fin_3 = time.time()
        tiempo_3 = (fin_3 - inicio_3)*1e6

        print(f'\n Matrices {i}x{i}:\n')
        print("Tiempo de ejecución de Laplace:", tiempo_1, "microsegundos")
        print("Tiempo de ejecución de Gauss:", tiempo_2, "microsegundos")
        print("Tiempo de ejecución de funcion predeterminada:", tiempo_3, "microsegundos")

In [ ]:
comparacion_tiempos()


 Matrices 2x2:

Tiempo de ejecución de Laplace: 0.0 microsegundos
Tiempo de ejecución de Gauss: 0.0 microsegundos
Tiempo de ejecución de funcion predeterminada: 0.0 microsegundos

 Matrices 3x3:

Tiempo de ejecución de Laplace: 0.0 microsegundos
Tiempo de ejecución de Gauss: 0.0 microsegundos
Tiempo de ejecución de funcion predeterminada: 0.0 microsegundos

 Matrices 4x4:

Tiempo de ejecución de Laplace: 0.0 microsegundos
Tiempo de ejecución de Gauss: 0.0 microsegundos
Tiempo de ejecución de funcion predeterminada: 0.0 microsegundos

 Matrices 5x5:

Tiempo de ejecución de Laplace: 8005.857467651367 microsegundos
Tiempo de ejecución de Gauss: 0.0 microsegundos
Tiempo de ejecución de funcion predeterminada: 0.0 microsegundos

 Matrices 6x6:

Tiempo de ejecución de Laplace: 12289.047241210938 microsegundos
Tiempo de ejecución de Gauss: 0.0 microsegundos
Tiempo de ejecución de funcion predeterminada: 0.0 microsegundos

 Matrices 7x7:

Tiempo de ejecución de Laplace: 81802.1297454834 micro

Observamos como el tiempo de ejecución en para Laplace es mucho mayor que en los otros dos casos, que son prácticamente inmediatos. Esto es consecuencia de lo comentado en el apartado de la complejidad computacional.

### Ejercicio 2

#### Apartado a)

Se define en primer lugar la función *grad_func*, que tiene dos parámetros: el punto donde se evalua el gradiente y la función cuyo gradiente hay que calcular (o derivada en el caso unidimensional). Aquí es donde se crean los gradientes de las funciones f y g del enunciado.


In [ ]:
def grad_func(x, funcion):

    if funcion == 'f':
        y = 12*(x**3)+12*(x**2)-24*x
    else:
        y = np.array([2*x[0]+3*x[1], 3*(x[1]**2)+3*x[0]])
    return y

Diseñamos ahora la funcion que implementa el método de descenso de gradiente con los parámetros descritos más la función:

In [ ]:
def descenso_grad(x, ratio, tol, maxit, funcion):

    grad_norm = linalg.norm(grad_func(x, funcion))
    cont = 0 #contador de iteraciones

    while grad_norm >= tol and cont <= maxit:
        grad = grad_func(x, funcion)
        x = x - ratio*grad
        grad_norm = linalg.norm(grad)
        cont += 1

    return x

#### Apartado b)

1) Se aplica el método sobre $f(x)$ con los parámetros indicados:

In [ ]:
descenso_grad(3, 0.001, 1e-12, 1e5, 'f')

1.0000000000000266

2) Se aplica el método sobre $f(x)$ con los parámetros indicados:

In [ ]:
descenso_grad(3, 0.01, 1e-12, 1e5, 'f')

-1.9999999999999967

3) Como $f'(x) = 12x(x^2+x-2)$, se deduce fácilmente que los ceros de la derivada son x=0, 1, -2. Evaluando dichos puntos en la segunda derivada

$$f''(x) = 36x^2 +24x-24,$$

se obtiene que:
- $f''(0) = -24 < 0    ->0$ es un máximo.
- $f''(1) = 36 > 0    -> 1$  es un mínimo.
- $f''(-2) = 72 > 0    ->2$ es un mínimo.

Aunque el punto inicial era el mismo para los apartados 1 y 2, el ratio era distinto. El ratio marca cuanto nos movemos en la dirección de máximo descenso. Por tanto, la elección de un ratio relativamente grande puede llegar a tener como consecuencia que converjamos a un mínimo que no es el más cercano al punto de partida. Incluso si el ratio es demasiado grande, el método puede llegar a no converger como se verá en el siguiente apartado.

4) Aplicamos el método sobre $f(x)$ con los parámetros indicados, pero al ejecutar nos da un error de que el valor es demasiado grande de calcular.

In [ ]:
descenso_grad(3, 0.1, 1e-12, 1e5, 'f') # Si se ejecuta da error

OverflowError: (34, 'Result too large')

El motivo de que de error es que se está tomando un ratio demasiado grande.

Lo que esta ocurriendo es que en cada iteración se desplace en el sentido del mínimo, se esta desplazando 'de más' y por tanto el método no converge. Es decir, los puntos $x_n$ calculados van apareciendo a la izquierda y derecha del mínimo alternadamente  pero cada vez más lejos del mínimo.

5. Si aplicamos ahora el método con punto inicial 0:

In [ ]:
descenso_grad(0, 0.001, 1e-12, 1e5, 'f')

0

Esto no es un resultado deseable puesto que es un máximo. Al ser un máximo la derivada se anula y nuestro método no  funciona y en consecuencia todos los $x_n$ sucesivos serán también cero, luego el método converge al mismo punto de inicio si es un máximo.

#### Apartado c)

1) Se aplica el método con los parámetros indicados, se obtiene el punto donde se alcanza el mínimo, como se verá en el último apartado:

In [ ]:
descenso_grad(np.array([-1,1]), 0.01, 1e-12, 1e5, 'g')

array([-2.25,  1.5 ])

2. Si se parte ahora del punto origen, se obtiene un resultado no deseable pues (0,0) no es un mínimo como se verá en el siguiente apartado.

In [ ]:
descenso_grad(np.array([0,0]), 0.01, 1e-12, 1e5, 'g')

array([0, 0])

3. En el primer apartado a) obtuvimos el gradiente de $g(x,y)$. Igualando el gradiente al vector nulo se obtienen los puntos críticos (resolviendo un sistema de dos ecuaciones con dos incógnitas), que son: $a=(0,0)$ y $b=(-2.25, 1.5)$